### ISSUE TITLE
reflect in documentation - deltachat-android submodule does not automatically update to newest master commit

### ISSUE THREAD

**Issue by pjvleeuwen** [deltachat/deltachat-android#2616]
On a fresh checkout of `deltachat-android`
with git modules also checked out
after running `nix develop`
the command `scripts/ndk-make.sh`
results in error:

```
[armeabi-v7a] Compile++ thumb: native-utils <= org_thoughtcrime_securesms_util_FileUtils.cpp
[armeabi-v7a] Compile thumb  : native-utils <= dc_wrapper.c
jni/dc_wrapper.c:1382:12: warning: implicit declaration of function 'dc_chat_is_protection_broken' is invalid in C99 [-Wimplicit-function-declaration]
    return dc_chat_is_protection_broken(get_dc_chat(env, obj))!=0;
           ^
1 warning generated.
[armeabi-v7a] SharedLibrary  : libnative-utils.so
ld: error: undefined symbol: dc_chat_is_protection_broken
>>> referenced by dc_wrapper.c:1382 (jni/dc_wrapper.c:1382)
>>>               lto.tmp:(Java_com_b44t_messenger_DcChat_isProtectionBroken)
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [/nix/store/vii7hnfm7lvg23qd5mfyki5gh4i810sa-ndk-23-2-8568313-23.2.8568313/build/core/build-binary.mk:716: obj/local/armeabi-v7a/libnative-utils.so] Error 1
```

@Hocuri this seems to be introduced today by this line:
https://github.com/deltachat/deltachat-android/commit/11a2cfbff7e22dfd8379bc641587f71926b78140#diff-e1cf1748c90a2c0e7ac38b9944678b342afee021012e9db17b698f86f86ef031

I was able to build the previous commit from yesterday just fine:
https://github.com/deltachat/deltachat-android/commit/df5537f0f6dd6e938b56cc7795e751b3cd185bf5

**Comment by pjvleeuwen** [comment]
ah, I guess that is why we have a origin/stable branch :) I will work from that

**Comment by Hocuri** [comment]
Yes, for working from the `master` branch, you'll need to `cd jni/deltachat-core-rust; git pull`.

I'm wondering if we should either document this or update the submodule on the `master` branch - IIRC we're usually only updating it on releases.

**Comment by pjvleeuwen** [comment]
@Hocuri

If I understand correctly : a specific commit of deltachat-android points to the specific commit of deltachat-core-rust with which it is compatible. At least for the commits that signify a release.

I am not sure if it is possible to point always to 'master' of the depending repo, but even if it is, it might not be desirable: you loose the information which commit is compatible with which commit from the other repo. This is not a problem when working with the master commit of each, but if you take an older commit of deltachat-android it is worthwhile to know which commit of deltachat-core-rust it is compatible with (chance is the older deltachat-android commit is _not_ compatible with 'master' of deltachat-core-rust).

The suggestion to always update the reference for every commit on the master branch that is affected is a nice idea, but might require significant more effort, discipline and risk of forgetting. So it might be good to document how to work with commits in between releases either way: as it allows for others to be able to always build master even in between commits.

At this point it is a bit unclear to me what we should instruct the others to do. Running `git pull --recurse-submodules` is only relevant if we indeed keep this reference up to date for each commit. Same goes for your `cd` + `pull` suggestion I think. [This suggestion](https://stackoverflow.com/a/14962718) might fix it, but I am not sure how to test it. What we would have to document would be `git submodule foreach git fetch ; git submodule foreach git checkout master; git submodule foreach git merge origin/master`. Not sure if this could be done easier / shorter.

A good place to mention this might be as an extra bullet point in this list: https://github.com/deltachat/deltachat-android/blob/11a2cfbff7e22dfd8379bc641587f71926b78140/README.md?plain=1#L19

**Comment by pjvleeuwen** [comment]
I did some experimentation to see what behavior submodules gives on updating to the latest master branch of the included external repo.

<details><summary>click to expand experimentation</summary><p>

> First create two bare repositories stubbing for the online github repositories. Create a working copy of both repositories and push an initial commit on both.
> 
> ```
> vk4omenwsl@vk4omen:~$ docker run -ti --rm bitnami/git
> root@cbc518542e09:/# git config --global user.email some@local.test
> root@cbc518542e09:/# git config --global user.name 'Some Test'
> 
> root@cbc518542e09:/# mkdir /tmp/github-core.git ; git -C /tmp/github-core.git init -b main --bare
> Initialized empty Git repository in /tmp/github-core.git/
> 
> root@cbc518542e09:/# mkdir /tmp/github-android.git ; git -C /tmp/github-android.git init -b main --bare
> Initialized empty Git repository in /tmp/github-android.git/
> 
> root@cbc518542e09:/# git clone /tmp/github-core.git /tmp/local-core
> Cloning into '/tmp/local-core'...
> warning: You appear to have cloned an empty repository.
> done.
> 
> root@cbc518542e09:/# git clone /tmp/github-android.git /tmp/local-android
> Cloning into '/tmp/local-android'...
> warning: You appear to have cloned an empty repository.
> done.
> 
> root@cbc518542e09:/# cd /tmp/local-core ; echo 'core 1' > foo.txt ; git add . ; git commit -m 'initial commit on core' ; git push
> [main (root-commit) 0943034] initial commit on core
>  1 file changed, 1 insertion(+)
>  create mode 100644 foo.txt
> Enumerating objects: 3, done.
> Counting objects: 100% (3/3), done.
> Writing objects: 100% (3/3), 221 bytes | 221.00 KiB/s, done.
> Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
> To /tmp/github-core.git
>  * [new branch]      main -> main
> 
> root@cbc518542e09:/tmp/local-core# cd /tmp/local-android ; echo 'android 1' > foo.txt ; git add . ; git commit -m 'initial commit on android' ; git push
> [main (root-commit) f73851e] initial commit on android
>  1 file changed, 1 insertion(+)
>  create mode 100644 foo.txt
> Enumerating objects: 3, done.
> Counting objects: 100% (3/3), done.
> Writing objects: 100% (3/3), 224 bytes | 224.00 KiB/s, done.
> Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
> To /tmp/github-android.git
>  * [new branch]      main -> main
> ```
> 
> Embed the `core` repo as submodule in the android repo. When a change is pushed to `core` it is not directly visible in the submodule directory. Also when doing a `git pull --recurse-submodules` the submodule does not pull this update on the remote branch.
> 
> ```
> root@cbc518542e09:/tmp/local-android# git config --global protocol.file.allow always
> 
> root@cbc518542e09:/tmp/local-android# git submodule add /tmp/github-core.git
> Cloning into '/tmp/local-android/github-core'...
> done.
> 
> root@cbc518542e09:/tmp/local-android# cat /tmp/local-android/github-core/foo.txt
> core 1
> 
> root@cbc518542e09:/tmp/local-android# cd /tmp/local-core ; echo 'core 2' > foo.txt ; git add . ; git commit -m 'update on core' ; git push
> [main 0c3e25a] update on core
>  1 file changed, 1 insertion(+), 1 deletion(-)
> Enumerating objects: 5, done.
> Counting objects: 100% (5/5), done.
> Writing objects: 100% (3/3), 243 bytes | 243.00 KiB/s, done.
> Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
> To /tmp/github-core.git
>    0943034..0c3e25a  main -> main
> 
> root@cbc518542e09:/tmp/local-core# cd /tmp/local-android/ ; git pull --recurse-submodules
> Fetching submodule github-core
> From /tmp/github-core
>    0943034..0c3e25a  main       -> origin/main
> Already up to date.
> 
> root@cbc518542e09:/tmp/local-android# cat /tmp/local-android/github-core/foo.txt
> core 1
> ```
> 
> After doing this whole submodule fetch, checkout and merge thing the submodule does show the latest commit on the remote branch. But after again doing a `git pull --recurse-submodules` the submodule directory is restored to the initial commit.
> 
> ```
> root@cbc518542e09:/tmp/local-android# git submodule foreach git fetch ; git submodule foreach git checkout main; git submodule foreach git merge origin/main
> Entering 'github-core'
> Entering 'github-core'
> Already on 'main'
> Your branch is behind 'origin/main' by 1 commit, and can be fast-forwarded.
>   (use "git pull" to update your local branch)
> Entering 'github-core'
> Updating 0943034..0c3e25a
> Fast-forward
>  foo.txt | 2 +-
>  1 file changed, 1 insertion(+), 1 deletion(-)
> 
> root@cbc518542e09:/tmp/local-android# cat /tmp/local-android/github-core/foo.txt
> core 2
> 
> root@cbc518542e09:/tmp/local-android# cd /tmp/local-core ; echo 'core 3' > foo.txt ; git add . ; git commit -m 'next update on core' ; git push
> [main cc242dc] next update on core
>  1 file changed, 1 insertion(+), 1 deletion(-)
> Enumerating objects: 5, done.
> Counting objects: 100% (5/5), done.
> Writing objects: 100% (3/3), 249 bytes | 249.00 KiB/s, done.
> Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
> To /tmp/github-core.git
>    0c3e25a..cc242dc  main -> main
> 
> root@cbc518542e09:/tmp/local-core# cd /tmp/local-android/ ; git pull --recurse-submodules
> Fetching submodule github-core
> From /tmp/github-core
>    0c3e25a..cc242dc  main       -> origin/main
> Already up to date.
> Submodule path 'github-core': checked out '09430341379aa8cda633efcd1bc80b1daa16ec45'
> 
> root@cbc518542e09:/tmp/local-android# cat /tmp/local-android/github-core/foo.txt
> core 1
> ```
> 
> So the trick seems to be to do the submodule fetch/checkout/main routine and not do the submodule pull.
> 
> ```
> root@cbc518542e09:/tmp/local-android# git submodule foreach git fetch ; git submodule foreach git checkout main; git submodule foreach git merge origin/main
> Entering 'github-core'
> Entering 'github-core'
> Previous HEAD position was 0943034 initial com mit on core
> Switched to branch 'main'
> Your branch is behind 'origin/main' by 1 commit, and can be fast-forwarded.
>   (use "git pull" to update your local branch)
> Entering 'github-core'
> Updating 0c3e25a..cc242dc
> Fast-forward
>  foo.txt | 2 +-
>  1 file changed, 1 insertion(+), 1 deletion(-)
> root@cbc518542e09:/tmp/local-android# cat /tmp/local-android/github-core/foo.txt
> core 3
> ```

</p></details>

TL;DR :

- a specific commit of `deltachat-android` points to the specific commit of `deltachat-core-rust`
- doing a pull recursing on the submodules does not pull any newer commit of `deltachat-core-rust` but restores to the above mentioned specific commit
- doing a submodule fetch/checkout/main routine does allow you to temporarily put the latest commit on the place of the submodule

I will propose an update to the documentation for review.

**Comment by pjvleeuwen** [comment]
PS: if I am off-track on this whole submodule thing, please let me know. I have almost no experience using them.

**Comment by pjvleeuwen** [comment]
It seems that the pull request build also [fails on this](https://github.com/deltachat/deltachat-android/actions/runs/5569569715/jobs/10173014887?pr=2618#step:9:1962)... this seems to be an argument that updating the submodule reference (also in between releases) is a must when breaking- / api changes are introduced.

**Comment by Hocuri** [comment]
Yes, I'm also convinced that we should update the submodule by now: https://github.com/deltachat/deltachat-android/pull/2619
